lib/softoken/kbkdf.c

```diff
@@ -0,0 +1,849 @@
+    if (session == NULL) {
+        ret = CKR_HOST_MEMORY;
+        goto done;
+    }
+
+    ret = sftk_handleObject(key, session);
+    *(derived_key->phKey) = key->handle;
+    /* TODO: Figure out why we still set phKey when ret != CKR_OK? */
+
+done:
+    /* NULL guard necessary as sftk_FreeObject(...) doesn't handle a NULL
+     * parameter. */
+    if (key != NULL) {
+        sftk_FreeObject(key);
+    }
+
+    /* Doesn't do anything. */
+    sftk_FreeSession(session);
+
+    return ret;
+}
+
+CK_RV kbkdf_SaveKeys(CK_SESSION_HANDLE hSession, CK_SP800_108_KDF_PARAMS_PTR params, uint8_t *output_buffer, size_t buffer_len, SFTKObject *ret_key, CK_ULONG keySize, size_t *key_index, uint8_t **key_buffer, size_t *key_buffer_index) {
+    CK_RV ret;
+
+    /* TODO: Simplify this method call. */
+
+    PR_ASSERT(output_buffer != NULL && key_index != NULL && key_buffer != NULL && key_buffer_index != NULL);
+
+    /* Two cases here:
+     *
+     *  - key_index == 0  --> we need to place key material into the main key
+     *                        passed directly to C_Derive.
+     *  - key_index > 0   --> we need to place key material into additional keys
+     *                        from params.
+     *
+     * Note that by PKCS#11 v3.0, Section 2.42.6 "Deriving Additional Keys",
+     * we _cannot_ place data from this PRF invocation into more than one
+     * output key. This means that any additional key material (after filling
+     * the current key) is simply discarded.
+     */
+
+    if (*key_index == 0) {
+        /* When key_index is at index 0, that means that we still need to
+         * place key material into the main key, or we've been passed a
+         * derivative key to place data into. There are two main cases here:
+         *
+         *  - "we have enough data for this key" --> Place it into the key.
+         *  - "we don't have enough data"        --> Store it (until the next
+         *                                           PRF invocation) in
+         *                                           *key_buffer.
+         */
+        if (*key_buffer == NULL) {
+            *key_buffer = calloc(keySize, sizeof(uint8_t));
+            PR_ASSERT(*key_buffer_index == 0);
+        }
+
+        if ((buffer_len + *key_buffer_index) >= keySize) {
+            /* Copy the data we need just for _this_ key. */
+            size_t copy_amount = keySize - *key_buffer_index;
+            PR_ASSERT(keySize > *key_buffer_index);
+            PORT_Memcpy((*key_buffer) + (*key_buffer_index), output_buffer, copy_amount);
+
+            PR_ASSERT(ret_key != NULL);
+
+            /* Save this key. */
+            ret = kbkdf_SaveKey(ret_key, *key_buffer, keySize);
+            if (ret != CKR_OK) {
+                return ret;
+            }
+
+            /* Clear and free the buffer, resetting back to the initial
+             * buffer state. */
+            PORT_Memset(*key_buffer, 0, keySize);
+            free(*key_buffer);
```

Comment:
Good, We need to clear all our CSPs for 1) good crypto hygeine, and 2) FIPS.

Reply:
I've switched these to ZFree, which should both zero and free.